File: api/src/main/java/org/apache/gravitino/function/FunctionChange.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.function;

import com.google.common.base.Preconditions;
import java.util.Arrays;
import java.util.Objects;
import org.apache.commons.lang3.StringUtils;
import org.apache.gravitino.annotation.Evolving;

/** Represents a change that can be applied to a function. */
@Evolving
public interface FunctionChange {
  /** An empty array of parameters. */
  FunctionParam[] EMPTY_PARAMS = new FunctionParam[0];

  /**
   * Create a {@link FunctionChange} to update the comment of a function.
   *
   * @param newComment The new comment value.
   * @return The change instance.
   */
  static FunctionChange updateComment(String newComment) {
    return new UpdateComment(newComment);
  }

  /**
   * Create a {@link FunctionChange} to add a new definition (overload) to a function.
   *
   * @param definition The new definition to add.
   * @return The change instance.
   */
  static FunctionChange addDefinition(FunctionDefinition definition) {
    return new AddDefinition(definition);
  }

  /**
   * Create a {@link FunctionChange} to remove an existing definition (overload) from a function.
   *
   * @param parameters The parameters that identify the definition to remove.
   * @return The change instance.
   */
  static FunctionChange removeDefinition(FunctionParam[] parameters) {
    return new RemoveDefinition(parameters);
  }

  /**
   * Create a {@link FunctionChange} to add an implementation to a specific definition.
   *
   * @param parameters The parameters that identify the definition to update.
   * @param implementation The implementation to add.
   * @return The change instance.
   */
  static FunctionChange addImpl(FunctionParam[] parameters, FunctionImpl implementation) {
    return new AddImpl(parameters, implementation);
  }

  /**
   * Create a {@link FunctionChange} to update an implementation for a specific definition and
   * runtime.
   *
   * @param parameters The parameters that identify the definition to update.
   * @param runtime The runtime that identifies the implementation to replace.
   * @param implementation The new implementation.
   * @return The change instance.
   */
  static FunctionChange updateImpl(
      FunctionParam[] parameters, FunctionImpl.RuntimeType runtime, FunctionImpl implementation) {
    return new UpdateImpl(parameters, runtime, implementation);
  }

  /**
   * Create a {@link FunctionChange} to remove an implementation for a specific definition and
   * runtime.
   *
   * @param parameters The parameters that identify the definition to update.
   * @param runtime The runtime that identifies the implementation to remove.
   * @return The change instance.
   */
  static FunctionChange removeImpl(FunctionParam[] parameters, FunctionImpl.RuntimeType runtime) {
    return new RemoveImpl(parameters, runtime);
  }

  /** A {@link FunctionChange} to update the comment of a function. */
  final class UpdateComment implements FunctionChange {
    private final String newComment;

    UpdateComment(String newComment) {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(newComment), "New comment cannot be null or empty");
      this.newComment = newComment;
    }

    /**
     * @return The new comment of the function.
     */
    public String newComment() {
      return newComment;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof UpdateComment)) {
        return false;
      }
      UpdateComment that = (UpdateComment) obj;
      return Objects.equals(newComment, that.newComment);
    }

    @Override
    public int hashCode() {
      return Objects.hash(newComment);
    }

    @Override
    public String toString() {
      return "UpdateComment{newComment='" + newComment + "'}";
    }
  }

  /** A {@link FunctionChange} to add a new definition to a function. */
  final class AddDefinition implements FunctionChange {
    private final FunctionDefinition definition;

    AddDefinition(FunctionDefinition definition) {
      Preconditions.checkArgument(definition != null, "Definition cannot be null");
      this.definition = definition;
    }

    /**
     * @return The definition to add.
     */
    public FunctionDefinition definition() {
      return definition;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof AddDefinition)) {
        return false;
      }
      AddDefinition that = (AddDefinition) obj;
      return Objects.equals(definition, that.definition);
    }

    @Override
    public int hashCode() {
      return Objects.hash(definition);
    }

    @Override
    public String toString() {
      return "AddDefinition{definition=" + definition + '}';
    }
  }

  /** A {@link FunctionChange} to remove an existing definition from a function. */
  final class RemoveDefinition implements FunctionChange {
    private final FunctionParam[] parameters;

    RemoveDefinition(FunctionParam[] parameters) {
      Preconditions.checkArgument(parameters != null, "Parameters cannot be null");
      this.parameters = Arrays.copyOf(parameters, parameters.length);
    }

    /**
     * @return The parameters that identify the definition to remove.
     */
    public FunctionParam[] parameters() {
      return parameters.length == 0 ? EMPTY_PARAMS : Arrays.copyOf(parameters, parameters.length);
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof RemoveDefinition)) {
        return false;
      }
      RemoveDefinition that = (RemoveDefinition) obj;
      return Arrays.equals(parameters, that.parameters);
    }

    @Override
    public int hashCode() {
      return Arrays.hashCode(parameters);
    }

    @Override
    public String toString() {
      return "RemoveDefinition{parameters=" + Arrays.toString(parameters) + '}';
    }
  }

  /** A {@link FunctionChange} to add an implementation to a definition. */
  final class AddImpl implements FunctionChange {
    private final FunctionParam[] parameters;
    private final FunctionImpl implementation;

    AddImpl(FunctionParam[] parameters, FunctionImpl implementation) {
      Preconditions.checkArgument(parameters != null, "Parameters cannot be null");
      this.parameters = Arrays.copyOf(parameters, parameters.length);
      Preconditions.checkArgument(implementation != null, "Implementation cannot be null");
      this.implementation = implementation;
    }

    /**
     * @return The parameters that identify the definition to update.
     */
    public FunctionParam[] parameters() {
      return parameters.length == 0 ? EMPTY_PARAMS : Arrays.copyOf(parameters, parameters.length);
    }

    /**
     * @return The implementation to add.
     */
    public FunctionImpl implementation() {
      return implementation;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof AddImpl)) {
        return false;
      }
      AddImpl that = (AddImpl) obj;
      return Arrays.equals(parameters, that.parameters)
          && Objects.equals(implementation, that.implementation);
    }

    @Override
    public int hashCode() {
      int result = Arrays.hashCode(parameters);
      result = 31 * result + Objects.hashCode(implementation);
      return result;
    }

    @Override
    public String toString() {
      return "AddImpl{parameters="
          + Arrays.toString(parameters)
          + ", implementation="
          + implementation
          + '}';
    }
  }

  /**
   * A {@link FunctionChange} to replace an implementation (identified by runtime) for a specific
   * definition.
   */
  final class UpdateImpl implements FunctionChange {
    private final FunctionParam[] parameters;
    private final FunctionImpl.RuntimeType runtime;
    private final FunctionImpl implementation;

    UpdateImpl(
        FunctionParam[] parameters, FunctionImpl.RuntimeType runtime, FunctionImpl implementation) {
      Preconditions.checkArgument(parameters != null, "Parameters cannot be null");
      this.parameters = Arrays.copyOf(parameters, parameters.length);
      Preconditions.checkArgument(runtime != null, "Runtime cannot be null");
      this.runtime = runtime;
      Preconditions.checkArgument(implementation != null, "Implementation cannot be null");
      this.implementation = implementation;
      Preconditions.checkArgument(
          runtime == implementation.runtime(),
          "Runtime of implementation must match the runtime being updated");
    }

    /**
     * @return The parameters that identify the definition to update.
     */
    public FunctionParam[] parameters() {
      return parameters.length == 0 ? EMPTY_PARAMS : Arrays.copyOf(parameters, parameters.length);
    }

    /**
     * @return The runtime that identifies the implementation to replace.
     */
    public FunctionImpl.RuntimeType runtime() {
      return runtime;
    }

    /**
     * @return The new implementation.
     */
    public FunctionImpl implementation() {
      return implementation;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof UpdateImpl)) {
        return false;
      }
      UpdateImpl that = (UpdateImpl) obj;
      return Arrays.equals(parameters, that.parameters)
          && runtime == that.runtime
          && Objects.equals(implementation, that.implementation);
    }

    @Override
    public int hashCode() {
      int result = Arrays.hashCode(parameters);
      result = 31 * result + Objects.hash(runtime, implementation);
      return result;
    }

    @Override
    public String toString() {
      return "UpdateImpl{parameters="
          + Arrays.toString(parameters)
          + ", runtime="
          + runtime
          + ", implementation="
          + implementation
          + '}';
    }
  }

  /** A {@link FunctionChange} to remove an implementation for a specific runtime. */
  final class RemoveImpl implements FunctionChange {
    private final FunctionParam[] parameters;
    private final FunctionImpl.RuntimeType runtime;

    RemoveImpl(FunctionParam[] parameters, FunctionImpl.RuntimeType runtime) {
      Preconditions.checkArgument(parameters != null, "Parameters cannot be null");
      this.parameters = Arrays.copyOf(parameters, parameters.length);
      Preconditions.checkArgument(runtime != null, "Runtime cannot be null");
      this.runtime = runtime;
    }

    /**
     * @return The parameters that identify the definition to update.
     */
    public FunctionParam[] parameters() {
      return parameters.length == 0 ? EMPTY_PARAMS : Arrays.copyOf(parameters, parameters.length);
    }

    /**
     * @return The runtime that identifies the implementation to remove.
     */
    public FunctionImpl.RuntimeType runtime() {
      return runtime;
    }

    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof RemoveImpl)) {
        return false;
      }
      RemoveImpl that = (RemoveImpl) obj;
      return Arrays.equals(parameters, that.parameters) && runtime == that.runtime;
    }

    @Override
    public int hashCode() {
      int result = Arrays.hashCode(parameters);
      result = 31 * result + Objects.hashCode(runtime);
      return result;
    }

    @Override
    public String toString() {
      return "RemoveImpl{parameters=" + Arrays.toString(parameters) + ", runtime=" + runtime + '}';
    }
  }
}

File: clients/client-python/gravitino/api/function/function_change.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from abc import ABC
from typing import List

from gravitino.api.function.function_definition import FunctionDefinition
from gravitino.api.function.function_impl import FunctionImpl
from gravitino.api.function.function_param import FunctionParam


class FunctionChange(ABC):
    """Represents a change that can be applied to a function."""

    EMPTY_PARAMS: List[FunctionParam] = []
    """An empty list of parameters."""

    @staticmethod
    def update_comment(new_comment: str) -> "FunctionChange":
        """Create a FunctionChange to update the comment of a function.

        Args:
            new_comment: The new comment value.

        Returns:
            The change instance.
        """
        return UpdateComment(new_comment)

    @staticmethod
    def add_definition(definition: FunctionDefinition) -> "FunctionChange":
        """Create a FunctionChange to add a new definition (overload) to a function.

        Args:
            definition: The new definition to add.

        Returns:
            The change instance.
        """
        return AddDefinition(definition)

    @staticmethod
    def remove_definition(parameters: List[FunctionParam]) -> "FunctionChange":
        """Create a FunctionChange to remove an existing definition from a function.

        Args:
            parameters: The parameters that identify the definition to remove.

        Returns:
            The change instance.
        """
        return RemoveDefinition(parameters)

    @staticmethod
    def add_impl(
        parameters: List[FunctionParam], implementation: FunctionImpl
    ) -> "FunctionChange":
        """Create a FunctionChange to add an implementation to a specific definition.

        Args:
            parameters: The parameters that identify the definition to update.
            implementation: The implementation to add.

        Returns:
            The change instance.
        """
        return AddImpl(parameters, implementation)

    @staticmethod
    def update_impl(
        parameters: List[FunctionParam],
        runtime: FunctionImpl.RuntimeType,
        implementation: FunctionImpl,
    ) -> "FunctionChange":
        """Create a FunctionChange to update an implementation for a specific definition.

        Args:
            parameters: The parameters that identify the definition to update.
            runtime: The runtime that identifies the implementation to replace.
            implementation: The new implementation.

        Returns:
            The change instance.
        """
        return UpdateImpl(parameters, runtime, implementation)

    @staticmethod
    def remove_impl(
        parameters: List[FunctionParam], runtime: FunctionImpl.RuntimeType
    ) -> "FunctionChange":
        """Create a FunctionChange to remove an implementation for a specific definition.

        Args:
            parameters: The parameters that identify the definition to update.
            runtime: The runtime that identifies the implementation to remove.

        Returns:
            The change instance.
        """
        return RemoveImpl(parameters, runtime)


class UpdateComment(FunctionChange):
    """A FunctionChange to update the comment of a function."""

    def __init__(self, new_comment: str):
        if not new_comment or not new_comment.strip():
            raise ValueError("New comment cannot be null or empty")
        self._new_comment = new_comment

    def new_comment(self) -> str:
        """Returns the new comment of the function."""
        return self._new_comment

    def __eq__(self, other) -> bool:
        if not isinstance(other, UpdateComment):
            return False
        return self._new_comment == other._new_comment

    def __hash__(self) -> int:
        return hash(self._new_comment)

    def __repr__(self) -> str:
        return f"UpdateComment(newComment='{self._new_comment}')"


class AddDefinition(FunctionChange):
    """A FunctionChange to add a new definition to a function."""

    def __init__(self, definition: FunctionDefinition):
        if definition is None:
            raise ValueError("Definition cannot be null")
        self._definition = definition

    def definition(self) -> FunctionDefinition:
        """Returns the definition to add."""
        return self._definition

    def __eq__(self, other) -> bool:
        if not isinstance(other, AddDefinition):
            return False
        return self._definition == other._definition

    def __hash__(self) -> int:
        return hash(self._definition)

    def __repr__(self) -> str:
        return f"AddDefinition(definition={self._definition})"


class RemoveDefinition(FunctionChange):
    """A FunctionChange to remove an existing definition from a function."""

    def __init__(self, parameters: List[FunctionParam]):
        if parameters is None:
            raise ValueError("Parameters cannot be null")
        self._parameters = list(parameters)

    def parameters(self) -> List[FunctionParam]:
        """Returns the parameters that identify the definition to remove."""
        return (
            list(self._parameters) if self._parameters else FunctionChange.EMPTY_PARAMS
        )

    def __eq__(self, other) -> bool:
        if not isinstance(other, RemoveDefinition):
            return False
        return self._parameters == other._parameters

    def __hash__(self) -> int:
        return hash(tuple(self._parameters))

    def __repr__(self) -> str:
        return f"RemoveDefinition(parameters={self._parameters})"


class AddImpl(FunctionChange):
    """A FunctionChange to add an implementation to a definition."""

    def __init__(self, parameters: List[FunctionParam], implementation: FunctionImpl):
        if parameters is None:
            raise ValueError("Parameters cannot be null")
        self._parameters = list(parameters)
        if implementation is None:
            raise ValueError("Implementation cannot be null")
        self._implementation = implementation

    def parameters(self) -> List[FunctionParam]:
        """Returns the parameters that identify the definition to update."""
        return (
            list(self._parameters) if self._parameters else FunctionChange.EMPTY_PARAMS
        )

    def implementation(self) -> FunctionImpl:
        """Returns the implementation to add."""
        return self._implementation

    def __eq__(self, other) -> bool:
        if not isinstance(other, AddImpl):
            return False
        return (
            self._parameters == other._parameters
            and self._implementation == other._implementation
        )

    def __hash__(self) -> int:
        return hash((tuple(self._parameters), self._implementation))

    def __repr__(self) -> str:
        return (
            f"AddImpl(parameters={self._parameters}, "
            f"implementation={self._implementation})"
        )


class UpdateImpl(FunctionChange):
    """A FunctionChange to replace an implementation for a specific definition."""

    def __init__(
        self,
        parameters: List[FunctionParam],
        runtime: FunctionImpl.RuntimeType,
        implementation: FunctionImpl,
    ):
        if parameters is None:
            raise ValueError("Parameters cannot be null")
        self._parameters = list(parameters)
        if runtime is None:
            raise ValueError("Runtime cannot be null")
        self._runtime = runtime
        if implementation is None:
            raise ValueError("Implementation cannot be null")
        self._implementation = implementation
        if runtime != implementation.runtime():
            raise ValueError(
                "Runtime of implementation must match the runtime being updated"
            )

    def parameters(self) -> List[FunctionParam]:
        """Returns the parameters that identify the definition to update."""
        return (
            list(self._parameters) if self._parameters else FunctionChange.EMPTY_PARAMS
        )

    def runtime(self) -> FunctionImpl.RuntimeType:
        """Returns the runtime that identifies the implementation to replace."""
        return self._runtime

    def implementation(self) -> FunctionImpl:
        """Returns the new implementation."""
        return self._implementation

    def __eq__(self, other) -> bool:
        if not isinstance(other, UpdateImpl):
            return False
        return (
            self._parameters == other._parameters
            and self._runtime == other._runtime
            and self._implementation == other._implementation
        )

    def __hash__(self) -> int:
        return hash((tuple(self._parameters), self._runtime, self._implementation))

    def __repr__(self) -> str:
        return (
            f"UpdateImpl(parameters={self._parameters}, runtime={self._runtime}, "
            f"implementation={self._implementation})"
        )


class RemoveImpl(FunctionChange):
    """A FunctionChange to remove an implementation for a specific runtime."""

    def __init__(
        self, parameters: List[FunctionParam], runtime: FunctionImpl.RuntimeType
    ):
        if parameters is None:
            raise ValueError("Parameters cannot be null")
        self._parameters = list(parameters)
        if runtime is None:
            raise ValueError("Runtime cannot be null")
        self._runtime = runtime

    def parameters(self) -> List[FunctionParam]:
        """Returns the parameters that identify the definition to update."""
        return (
            list(self._parameters) if self._parameters else FunctionChange.EMPTY_PARAMS
        )

    def runtime(self) -> FunctionImpl.RuntimeType:
        """Returns the runtime that identifies the implementation to remove."""
        return self._runtime

    def __eq__(self, other) -> bool:
        if not isinstance(other, RemoveImpl):
            return False
        return self._parameters == other._parameters and self._runtime == other._runtime

    def __hash__(self) -> int:
        return hash((tuple(self._parameters), self._runtime))

    def __repr__(self) -> str:
        return f"RemoveImpl(parameters={self._parameters}, runtime={self._runtime})"
